### ISSUE TITLE
Add Italian GPM data + workflow

### ISSUE THREAD

**Issue by kmuehlbauer** [openradar/erad2024#112]
We did not make it before ERAD2024, but I've received 3 volumes of Italian radar data (thanks to Gianfranco) to extend our GPM Matching/Calibration.

@mgrover1  I'll check the data the next days and will contact you to put it on the pythia bucket.
@ghiggi It would be great if you could extend/add the processing for this data, once it's available.

**Comment by ghiggi** [comment]
Yes. I can do it as soon as the data are available. Which xradar reader should be used to read the Italian data? I can also provide a list of all GPM overpass over Italy for who want to extend the analysis. 
With @aladinor we might provide a tutorial for the Columbian radars too ;)

**Comment by kmuehlbauer** [comment]
> Which xradar reader should be used to read the Italian data?

It's in the datamet-format which @wolfidan implemented a reader just before ERAD2024.

```python
import xradar as xd
fname = "somename.tar.gz"
tree = xd.io.open_datamet_datatree(fname)
```

We'll have to wait until @mgrover1 is back in office. But you can already play with the data. Thanks @ghiggi.

**Comment by ghiggi** [comment]
@kmuehlbauer I had a quicklook this morning, but there is something strange in the GR data. 
This is the GPM DPR surface reflectivity measured at 06:30 UTC. The cross indicates the position of the GR
![image](https://github.com/user-attachments/assets/4bd3e5c5-967a-4a43-99a0-9154351ed71a)

If I look at the GR data, I see something that correlates to the shape of the storm at sweep higher than 6, but with reflectivities below 10 dBZ. 
 
![image](https://github.com/user-attachments/assets/2c3de7d1-eff0-4556-ac10-052444363e26)
![image](https://github.com/user-attachments/assets/89811b8d-77c8-4420-a790-ffea795b441d)

![image](https://github.com/user-attachments/assets/ea699e14-e867-47c6-92da-1d7cc7109a28)
![image](https://github.com/user-attachments/assets/7460580f-dd32-4858-b6ea-ed5b896cf403)

![image](https://github.com/user-attachments/assets/9a8d2a45-d449-44c1-befe-65766c55394b)
![image](https://github.com/user-attachments/assets/8d9d062c-60e7-4321-adad-aa5ba6652c79)

There is something going wrong with the decoding/scaling of the GR data I think @wolfidan ! If the `scale_factor` is defined as `1/current_scale_factor` in the reader I obtain this which looks more realistic to me ;) 

![image](https://github.com/user-attachments/assets/baf6a7e0-116f-439c-8b04-b0dff768c905)

**Comment by ghiggi** [comment]
@kmuehlbauer for the tutorial we will need to filter out the clutter and some additional preprocessing of GR data before doing the comparison. Do we use pyart or wradlib? Do you have already some default/template processing to use?

**Comment by kmuehlbauer** [comment]
@ghiggi The radar obviously operates in top-down mode. So higher elevations are scanned first.

You are referring to these [lines of code](https://github.com/openradar/xradar/blob/34a9e8234189f921b1586001c84bdaf8f80f4c48/xradar/io/backends/datamet.py#L166-L167)?

I'm not sure what the correct setting of `offset` and `slope` is, but if I use a negated offset (and keep the scale_factor), the output looks even more realistic. Can you confirm that @ghiggi?

For preprocessing examples I usually use clutter detection from wradlib to remove clutter and make some interpolation afterwards to fill the holes. But we can also use pyart for preprocessing. I've no fixed preprocessor pipeline, always build it from scratch for new radars.

**Comment by ghiggi** [comment]
I assumed C band radar, empirical correction of Tan for Ku to C, beamwidth of 1° ... plot some stuffs and I don't think it's a good overpass for assessing GR bias. Most volumes are above 35 dBZ, ground clutter, beam blockage, NUBF, ... I played a bit with filtering criteria but ...

![image](https://github.com/user-attachments/assets/5d91d263-c714-454b-88fd-b1a452f29d1e)
![image](https://github.com/user-attachments/assets/05831e4e-e612-4f0a-a4bb-c2ec5c3b4702)
 
![image](https://github.com/user-attachments/assets/ca8368c2-a20e-4e48-9275-5923086e62a1)

**Comment by ghiggi** [comment]
@kmuehlbauer I was looking more at [these lines](https://github.com/openradar/xradar/blob/34a9e8234189f921b1586001c84bdaf8f80f4c48/xradar/io/backends/datamet.py#L281).
I see `scale_factor` and `add_offset` added to the `attrs`, but I found these in the `encoding` once I open the datatree. This means xarray/cf decoding is doing something and moving from `attrs` to `encoding`. Are we applying scaling twice? [Here](https://github.com/openradar/xradar/blob/34a9e8234189f921b1586001c84bdaf8f80f4c48/xradar/io/backends/datamet.py#L182C9-L182C13) and using CF decoding? I can't have a look at it for the rest of the day ... I might be able to give a look tomorrow ...

**Comment by kmuehlbauer** [comment]
@ghiggi 

You can deal with this the following:

```python
import xarray as xr
swp = xr.open_dataset(fname, engine="datamet", group="sweep_11", decode_cf=False)
swp.DBZH.attrs["add_offset"] = swp.DBZH.attrs["add_offset"] * -1
swp = xr.decode_cf(swp)
```

`add_offset` and `scale_factor` are moved to the attributes to make them available for automatic decoding in xarray. This is what happens, but it looks like the offset needs to be the other way round.

Take your time, no rush.

**Comment by ghiggi** [comment]
Quickly tried and this is not correct:
```
import xarray as xr
swp = xr.open_dataset(fname, engine="datamet", group="sweep_11", decode_cf=False)
swp.DBZH.attrs["add_offset"] = swp.DBZH.attrs["add_offset"] * -1
swp = xr.decode_cf(swp)
```
If we open with `decode_cf=False` we get directly: 
![image](https://github.com/user-attachments/assets/e1c7061d-03ec-4d5c-a0a0-b991b5f0a612)
If we do 
```
    ds_gr["DBZH"].attrs["scale_factor"] = 1/ds_gr["DBZH"].attrs["scale_factor"] 
    ds_gr = xr.decode_cf(ds_gr)
```
we get this: 

![image](https://github.com/user-attachments/assets/6b7583b8-d5e5-4c3c-96fc-258dc244f8b1)

I think we are applying decoding twice ! 
By just reading with `decode_cf=False` and doing anything else I get this compared to GPM DPR
![image](https://github.com/user-attachments/assets/6dbc959e-8ab0-4a87-9d4c-b10d7802557d)

**Comment by kmuehlbauer** [comment]
Thanks @ghiggi, I'll have a deeper look later today.

**Comment by kmuehlbauer** [comment]
True @ghiggi, your hunch is correct! :clap: scale offset is applied twice, once in the reader and second time in xarray decoding.

https://github.com/openradar/xradar/blob/34a9e8234189f921b1586001c84bdaf8f80f4c48/xradar/io/backends/datamet.py#L182-L183

**Comment by kmuehlbauer** [comment]
@ghiggi FYI: https://github.com/openradar/xradar/pull/209

**Comment by mgrover1** [comment]
@kmuehlbauer - checking in here now that I am back in Chicago! Is there a link to the data to upload?

**Comment by kmuehlbauer** [comment]
ftp://ftp.meteo.uni-bonn.de/pub/radar/erad2024/gpm/vol-serano-202405210625-tar-gz_2024-09-16_0743/H-000-VOL-SERANO-202405210625.tar.gz

ftp://ftp.meteo.uni-bonn.de/pub/radar/erad2024/gpm/vol-serano-202405210625-tar-gz_2024-09-16_0743/H-000-VOL-SERANO-202405210630.tar.gz

ftp://ftp.meteo.uni-bonn.de/pub/radar/erad2024/gpm/vol-serano-202405210625-tar-gz_2024-09-16_0743/H-000-VOL-SERANO-202405210635.tar.gz

**Comment by mgrover1** [comment]
@ghiggi @kmuehlbauer - the data is now available in the pythia bucket under the /radar/erad2024/gpm_api directory:
- H-000-VOL-SERANO-202405210625.tar.gz
- H-000-VOL-SERANO-202405210630.tar.gz
- H-000-VOL-SERANO-202405210635.tar.gz

**Comment by kmuehlbauer** [comment]
@ghiggi I've released xradar v0.6.5 with the needed fix. When using the binder, we`d have to create an new PR to re-create the docker image with latest xradar.

**Comment by ghiggi** [comment]
Hi @kmuehlbauer ! I unfortunately didn't had time to advance on this. Next week I will be in Wurzburg at the EUMETSAT satellite conference ... so I will try to finalize this in two weeks ...

**Comment by ghiggi** [comment]
Hi @kmuehlbauer @mgrover1 ! I completely forgot about answering to this issue. 
When I went through the data, I discovered there was few matching footprints (after quality filtering) to make a meaningful comparison. 
In the past months I improved the volume matching routine and now I have scripts to assess the calibration of hundreds of ground radars during the period (2018-2024) through [radar_api](https://github.com/ghiggi/radar_api). If Gianfranco is interested, we could discuss about run it also for the Italian radars. 
BTW Kai I am coming to Bonn next week, so we might discuss further of it in person ;)

**Comment by mgrover1** [comment]
Great progress here @ghiggi - I really like all the progress you have made with radar_api 👍
